### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by dominikstraessle**
I corrected the installation instructions and wrote a backend documentation. I also added a table of contents to every documentation file.

While writing the documentation I had to implement the [GET users/](https://github.com/Simai00/jmp/blob/master/docs/api-v1.md#get-user) route, to know which steps are required to implement a new route.

As my english isn't excellent, please read every documentation very precise and correct the mistakes. I think it's easier to pull the branch than correcting directly in the pull request.

### CODE REVIEW

File: docker/db/01_setup.sql [hunk 1]
```diff
@@ -111,15 +111,15 @@ DEFAULT CHARACTER SET = utf8;
...
-  `id` INT(11) NOT NULL AUTO_INCREMENT,
-  `username` VARCHAR(101) NOT NULL UNIQUE,
-  `lastname` VARCHAR(50),
-  `firstname` VARCHAR(50),
-  `email` VARCHAR(255),
-  `password` VARCHAR(255) NOT NULL,
-  `password_change` TINYINT DEFAULT 0 NOT NULL,
-  `is_admin` TINYINT DEFAULT 0 NOT NULL,
-  PRIMARY KEY (`id`))
+                                          `id`              INT(11)           NOT NULL AUTO_INCREMENT,
```
Comment (mtte) [inline docker/db/01_setup.sql]: Please fix the incorrect indentation and revert this change.

File: api/src/validation.php [hunk 2]
```diff
@@ -37,6 +37,9 @@
             'firstname' => v::optional(v::notEmpty()->noWhitespace()->length(1, 51)),
             'password' => v::notEmpty()->length(1, 256),
             'email' => v::optional(v::notEmpty()->length(1, 256)::email())
+        ],
+        'getUser' => [
```
Comment (mtte) [inline api/src/validation.php]: Remove this validation as the validation is already done by the expression matching of the route:
`/users/{id:[0-9]+`

File: api/src/routes.php [hunk 3]
```diff
@@ -46,4 +46,10 @@
         ->add(new AuthenticationMiddleware($container, \JMP\Utils\PermissionLevel::ADMIN))
         ->add($jwtMiddleware);
 
-});
\ No newline at end of file
+    $this->get('/users/{id:[0-9]+}', UsersController::class . ':getUser')
+        ->add(new ValidationErrorResponseBuilder())
+        ->add(new Validation($container['validation']['getUser']))
```
Comment (mtte) [inline api/src/routes.php]: Remove this validation as the validation is already done by the expression matching of the route:
`/users/{id:[0-9]+`

File: README.md [hunk 4]
```diff
@@ -1,56 +1,74 @@
...
+Build and start the docker containers
 ```bash
-composer install -d api
+docker-compose up -d
 ```
 
-### Docker
+Install all the dependencies using the composer of the app container
+````bash
+docker exec -it app composer install
```
Comment (mtte) [inline README.md]: I think this step is not necessary as the app container installs all composer dependencies when the app container is built.

File: docs/backend.md [hunk 5]
```diff
@@ -2,40 +2,285 @@
...
-```
-
-# Developing new route
+* [`src/`](../api/src) Source directory. Not public
+    * [`JMP/`](../api/src/JMP) Application directory
+        * [`Controllers/`](../api/src/JMP/Controllers) Controllers to handle requests and build the responses
+        * [`Middleware/`](../api/src/JMP/Middleware) Custom [Middlewares](https://www.slimframework.com/docs/v3/concepts/middleware.html)
+        * [``Models/``](../api/src/JMP/Models) Raw data objects with the functionality to convert to an array
+        * [``Services/``](../api/src/JMP/Services) Business logic and database communication
+        * [``Utils/``](../api/src/JMP/Utils) Useful classes
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
        * [``Utils/``](../api/src/JMP/Utils) Utility classes
```

File: docs/backend.md [hunk 6]
```diff
@@ -2,40 +2,285 @@
...
+If no token is supplied or if the supplied token is invalid, a 401 is returned.
+In case the enquirer just hasn't got the required permissions, a 403 is returned.
+### Login
+To receive a new token, the user have to call the [login route](api-v1.md#login).  
+If the username and the password are valid, a new token is generated with the following service class [Auth.php](../api/src/JMP/Services/Auth.php).
+The password is hashed using [`password_hash`](https://secure.php.net/manual/en/function.password-hash.php) with the `PASSWORD_DEFAULT` option before it is stored in the database.
+To verify a password at a login, [`password_verify`](https://secure.php.net/manual/en/function.password-verify.php) is used.
+### Registration
+Only administrators can register new users. To register a new user, the [register route](api-v1.md) has to be called.
+
```
Comment (mtte) [inline docs/backend.md]: Which route is meant? The create user route?

I'd remove the registration section as the documentation of endpoints should not be in this file.

File: docs/backend.md [hunk 7]
```diff
@@ -2,40 +2,285 @@
...
+    * [``settings.php``](../api/src/settings.php) Slim settings and required settings for the dependencies
+    * [``routes.php``](../api/src/routes.php) Register routes with their middlewares here
+* Public
+    * [``index.php``](../api/public/index.php) Application entry point
+
+
+## Authentication
+
+### Usage
+We use [JWT](https://jwt.io/) for authentication. The token must be submitted in the Authorization header using the Bearer schema.
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
We use [JWT](https://jwt.io/) for authentication. The token must be submitted in the authorization header using the bearer schema.  
```

File: docs/backend.md [hunk 8]
```diff
@@ -2,40 +2,285 @@
...
+```php
+$optional = methodWhichMayFail();
+if ($optional->isSuccess()) {
+    // success
+    $data = $optional->getData();
+} else {
+//failure
+}
+```
+**Vic versa:**
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
**Vice versa:**
```

File: docs/backend.md [hunk 9]
```diff
@@ -2,40 +2,285 @@
...
+if ($optional->isFailure()) {
+    // failure
+} else {
+    // success
+    $data = $optional->getData();
+}
+```
+
+## Array Conversion
+The response object of the slim framework offers a method called `withJson`. This method converts an assoc array to JSON.
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
The response object of the slim framework offers a method called `withJson`. This method converts an associative array to JSON.
```

File: docs/backend.md [hunk 10]
```diff
@@ -2,40 +2,285 @@
...
+```php
+return $response->withJson(Converter::convert($modelObject));
+```
+or
+```php
+return $response->withJson(Converter::convertArray($list));
+```
+
+## Type declarations & phpdoc
+Everywhere it is possible we use object oriented php. So every method signature have to use the [php type declarations](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration).
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
Everywhere it is possible we use object oriented php. So every method signature has to use the [php type declarations](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration).
```

File: docs/backend.md [hunk 11]
```diff
@@ -2,40 +2,285 @@
...
+If you try to call a method and pass arguments of another type as declared, then php will throw an [TypeError](https://secure.php.net/manual/en/class.typeerror.php).
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
If you try to call a method and pass arguments of another type as declared, then php will throw a [TypeError](https://secure.php.net/manual/en/class.typeerror.php).
```

File: docs/backend.md [hunk 12]
```diff
@@ -2,40 +2,285 @@
...
+
+Also, every method has to be documented with phpdoc. A short summary or description of the method and a documented signature is sufficient.
+
+**Example (from [UserService.php](../api/src/JMP/Services/UserService.php)):**
+```php
+/**
+ * Select a user by its id
+ * @param int $userId
+ * @return Optional containing a User on succeed
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
 * @return Optional containing a User if successfull
```

File: docs/backend.md [hunk 13]
```diff
@@ -2,40 +2,285 @@
...
+* Use the heredoc syntax:
+```php
+$sql = <<<SQL
+# SQL
+SQL;
+```
+* Write every sql keyword upper-case
+* Use the **`AS`** keyword to change column names from the sql underscore style to the camel case style. E.g. `SELECT user_id as userId ...`
+* Use the following for optional parameters: `WHERE (:optionalId IS NULL OR id = :optionalId)`. Examples in [EventService.php](../api/src/JMP/Services/EventService.php)
+* A specific service don't query tables to which the service doesn't belong unless it's a join
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
* A specific service doesn't query tables to which the service doesn't belong unless it's a join
```

File: docker/db/01_setup.sql (hunk 1 above)
Comment (mtte) [reply docker/db/01_setup.sql]: Fixed

File: docs/backend.md [hunk 14]
```diff
@@ -2,40 +2,271 @@
...
+    * [``settings.php``](../api/src/settings.php) Slim settings and required settings for the dependencies
+    * [``routes.php``](../api/src/routes.php) Register routes with their middlewares here
+* Public
+    * [``index.php``](../api/public/index.php) Application entry point
+
+
+## Authentication
+
+### Usage
+We use [JWT](https://jwt.io/) for authentication. The token must be submitted in the Authorization header using the Bearer schema.
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
We use [JWT](https://jwt.io/) for authentication. The token must be submitted in the authorization header using the bearer schema.
```

File: docs/backend.md [hunk 15]
```diff
@@ -2,40 +2,271 @@
...
+```php
+return $response->withJson(Converter::convert($modelObject));
+```
+or
+```php
+return $response->withJson(Converter::convertArray($list));
+```
+
+## Type declarations & phpdoc
+Everywhere it is possible we use object oriented php. So every method signature have to use the [php type declarations](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration). If you try to call a method and pass arguments of another type as declared, then php will throw an [TypeError](https://secure.php.net/manual/en/class.typeerror.php).
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
Everywhere it is possible we use object oriented php. So every method signature has to use the [php type declarations](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration). If you try to call a method and pass arguments of another type as declared, then php will throw a [TypeError](https://secure.php.net/manual/en/class.typeerror.php).
```

File: docs/backend.md [hunk 16]
```diff
@@ -2,40 +2,271 @@
...
+All validations are stored in [validation.php](../api/src/validation.php) and use the [RespectValidation](https://github.com/Respect/Validation) validation engine. The stored validations have to be submitted to the [validation middleware](https://github.com/DavidePastore/Slim-Validation) of the corresponding route in [routes.php](../api/src/routes.php).
+
+**Example:**
+```php
+->add(new Validation($container['validation']['getUser']))
+```
+See more at [Routes](#route).
+
+### Translations
+In case the automatic generated error messages doesn't fit your requirements or hold sensible data like a password, you can add translations as described in [Translate errors](https://github.com/DavidePastore/Slim-Validation#translate-errors).
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
In case the automatic generated error messages don't fit your requirements or hold sensible data like a password, you can add translations as described in [Translate errors](https://github.com/DavidePastore/Slim-Validation#translate-errors).
```

File: docs/backend.md [hunk 17]
```diff
@@ -2,40 +2,271 @@
...
+    // failure
+} else {
+    // success
+    $data = $optional->getData();
+}
+```
+
+## Array Conversion
+The response object of the slim framework offers a method called `withJson`. This method converts an assoc array to JSON.
+Because the php cast functionality doesn't comply our requirements to cast model objects to associative arrays, we use the following util and interface:
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
Because the php cast functionality doesn't comply with our requirements to cast model objects to associative arrays, we use the following util and interface:
```

File: docs/backend.md [hunk 18]
```diff
@@ -2,40 +2,271 @@
...
+ * @param int $userId
+ * @return Optional containing a User on succeed
+ */
+public function getUserByUserId(int $userId): Optional
+```
+
+## SQL
+The services handle the business logic as well as the whole database communication. At the moment there isn't a separated Persistence Layer with e.g. Data Access Objects (DAOs).
+
+That the services doesn't become a mess, all SQL have to be coded with the following rules:
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
That the services don't become a mess, all SQL statements have to be coded with the following rules:
```

File: docs/backend.md [hunk 19]
```diff
@@ -2,40 +2,271 @@
...
+Make sure you've read [Code Style](#code-style) before reading the following guide.
+
+## Api specification
+First of all the new route has to be documented in the [api-specification](api-v1.md).
 
 ## Service & Model
+Routes return JSON mostly containing one or a list of model objects. So you have to [create a new model](#create-a-new-model) if the required doesn't exist yet. The model is just a raw data object with the functionality to convert properly to an array.
+
+Then you need a service to communicate with the database and to handle the business logic.
+[Create a new service class](#create-a-new-service-class) if no appropriate exists, otherwise add the required method/s to the existing service class.
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
[Create a new service class](#create-a-new-service-class) if no appropriate exists yet, otherwise add the required method/s to the existing service class.
```

File: docs/backend.md [hunk 20]
```diff
@@ -2,40 +2,271 @@
...
+The service must
+* have a `ContainerInterface` constructor parameter
+* hold all dependencies as private attributes
+* have all dependencies set inside the constructor
+* be added to the slim container inside [dependencies.php](../api/src/dependencies.php)
+
+Check out already existing services as examples. [Services](../api/src/JMP/Services) 
+
+### Create a new model
+If the required model not already exists, you have to create a new one.
```
Comment (mtte) [inline docs/backend.md]:
```suggestion
If the required model doesn't already exist, you have to create a new one.
```

File: README.md (hunk 4 above)
Comment (dominikstraessle) [reply README.md]: Yes, it's necessary. I've just tested it again.

File: docs/backend.md (hunk 6 above)
Comment (dominikstraessle) [reply docs/backend.md]: It's also part of the authentication and a requirement of the project. So I think it's not wrong to link the registration of the api specification here.

It's the create user route, so I will update the link.
